### PR TITLE
Add GitHub environments support to custom workflows

### PR DESCRIPTION
**Pull request by klutchell**

### CODE REVIEW

**Comment by ab77** [comment]
@klutchell the only issue I have with this approach (if I understood it correctly), as that it seems to establish a naming lock-in between custom job names and GH environment names. The approach used previously, allows for flexibility to "deploy" custom matrix "A" to GH env. "B" (e.g.):

```
      custom_test_matrix: >
        A

      custom_matrices_includes: >
        [
          {
            "value": "A",
            "environment": "B"
          }
        ]
```

If you don't think this is a major issue, I only request that you keep the original commit in the PR for posterity, so when we merge, we have both commits for future reference.

**Comment by klutchell** [comment]
> @klutchell the only issue I have with this approach (if I understood it correctly), as that it seems to establish a naming lock-in between custom job names and GH environment names. The approach used previously, allows for flexibility to "deploy" custom matrix "A" to GH env. "B" (e.g.):
> 
> ```
>       custom_test_matrix: >
>         A
> 
>       custom_matrices_includes: >
>         [
>           {
>             "value": "A",
>             "environment": "B"
>           }
>         ]
> ```
> 
> If you don't think this is a major issue, I only request that you keep the original commit in the PR for posterity, so when we merge, we have both commits for future reference.

Good catch, and it was actually something I missed when updating the playground repo. Using `custom_environments` means we don't need to provide a `custom_*_matrix` at all if they are the same value. [Here](https://github.com/balena-io/environment-playground/pull/69/files) is the updated workflow, where we are providing only the environment and no longer need the matrices.

I also added a step to this PR to export the environment name to GITHUB_ENV in case the custom action needed to reference the environment by name.

Also, your original method for posterity and reference: https://github.com/product-os/flowzone/pull/964

I see the benefits of both methods, but I think this might be more useful if the matrix value is not an environment (which it wouldn't need to be anymore) but is something else, like a node version for example? Then we could have a matrix like this:

```yaml
matrix:
        value: ${{ fromJSON(needs.is_custom.outputs.custom_publish_matrix) }} # node18, node20
        os: ${{ fromJSON(inputs.custom_runs_on || format('[{0}]', inputs.runs_on)) }} # ubuntu-latest, macos-latest
        environment: ${{ fromJSON(needs.is_custom.outputs.custom_environments_matrix) }} # production, staging

    environment: ${{ matrix.environment }}
```

Results in 2 * 2 * 2 = 8 unique jobs

So it fully decouples the custom matrix value from the environment matrix, which I think is a more logical use case?

File: flowzone.yml [hunk 1]
```diff
@@ -2995,6 +3008,9 @@ jobs:
       matrix:
         value: ${{ fromJSON(needs.is_custom.outputs.custom_test_matrix) }}
         os: ${{ fromJSON(inputs.custom_runs_on || format('[{0}]', inputs.runs_on)) }}
+        environment: ${{ fromJSON(needs.is_custom.outputs.custom_environments_matrix) }}
+
+    environment: ${{ matrix.environment }}
```
[the remark this answers is not in the document]
Comment (klutchell) [reply flowzone.yml:3013]: I'll add the tests, good idea!

[the remark this answers is not in the document]
Comment (klutchell) [reply flowzone.yml:3098]: Added tests for vars and secrets only present in the `test` environment 👍

**Review by ab77 — APPROVED** [review]

**Pull request merged**